backend/src/main/kotlin/backend/graphql/BonusDataFetcher.kt

```
@@ -0,0 +1,267 @@
+        val teacher = getUsers(teacherId)
+        val subcategory = getSubcategory(subcategoryId)
+
+        val points = Points(
+            student = student,
+            teacher = teacher,
+            value = value,
+            subcategory = subcategory,
+            label = ""
+        )
+
+        val savedPoints = pointsRepository.save(points)
+
+        // Check all points the user has that are connected to some award
+        val userPoints = pointsRepository.findAllByStudent(student)
+
+        userPoints.forEach { point ->
+            val bonuses = bonusRepository.findByPoints(point)
+            bonuses.forEach { bonus ->
+                val award = bonus.award
+                if (award.awardType == AwardType.MULTIPLICATIVE) {
+                    // Recompute the value and update points entry
+                    val pointsInCategory = userPoints.filter { it.subcategory.category == award.category && bonusRepository.findByPoints(it).isEmpty() }
+                    val totalPointsValue = pointsInCategory.sumOf { it.value }
+                    point.value = (totalPointsValue * award.awardValue).toLong()
+                    pointsRepository.save(point)
+                } else {
+                    // Future logic can be added here
+                }
+            }
+        }
+
+        return savedPoints
+    }
+
+    private fun getUsers(userId: Long): Users {
+        return usersRepository.findById(userId)
+            .orElseThrow { IllegalArgumentException("Invalid user ID") }
+    }
+    private fun getSubcategory(subcategoryId: Long): Subcategories {
+        return subcategoriesRepository.findById(subcategoryId)
+            .orElseThrow { IllegalArgumentException("Invalid subcategory ID") }
+    }
+
+    private fun getChestHistory(chestHistoryId: Long): ChestHistory {
+        return chestHistoryRepository.findById(chestHistoryId)
+            .orElseThrow { IllegalArgumentException("Invalid chest history ID") }
+    }
```

I'm not sure if we should use those one liner functions. They hide error throws from main mutation function and in this way create hide possible endpoints of mutation. In this case I think it is better to keep all places where mutation can return something directly inside its body. Note that throwing error inside mutation has the same result as returning (and sending to hasura) message with error.